lib/archiver.js

```diff
@@ -0,0 +1,168 @@
+      await new Promise(resolve => archive.close(resolve))
+      delete this.archives[key]
+    }
+  }
+
+  async closeAllArchives () {
+    return Promise.all(Object.keys(this.archives).map(key =>
+      this.closeArchive(key)
+    ))
+  }
+
+  // internal
+  // =
+
+  _getArchiveFilesPath (key) {
+    return path.join(this.config.dir, 'archives', key.slice(0, 2), key.slice(2))
+  }
+
+  // load archive (inner) main load logic
+  async _loadArchiveInner (archivePath, key) {
+    // create the archive instance
+    var archive = hyperdrive(archivePath, key, {sparse: false})
+    archive.replicationStreams = [] // list of all active replication streams
+
+    // wait for ready
+    await new Promise((resolve, reject) => {
+      archive.ready(err => {
+        if (err) reject(err)
+        else resolve()
+      })
+    })
+
+    // join the swarm
+    this._joinSwarm(archive)
+
+    if (!archive.metadata.length) {
+      // wait to receive a first update
+      await new Promise((resolve, reject) => {
+        archive.metadata.update(err => {
+          if (err) reject(err)
+          else resolve()
+        })
+      })
+    }
+
+    // download the full metadata
+    await new Promise((resolve, reject) => {
+      archive.metadata.download({start: 0, end: archive.metadata.length}, err => {
+        if (err) reject(err)
+        else resolve()
+      })
+    })
+
+    return archive
+  }
+
+  _joinSwarm (archive) {
+    var key = datEncoding.toStr(archive.key)
+    var swarm = discoverySwarm(swarmDefaults({
+      hash: false,
+      utp: true,
+      tcp: true,
+      stream: (info) => {
+        var dkey = datEncoding.toStr(archive.discoveryKey)
+        var chan = dkey.slice(0, 6) + '..' + dkey.slice(-2)
+        var keyStrShort = key.slice(0, 6) + '..' + key.slice(-2)
+        debug('new connection chan=%s type=%s host=%s key=%s', chan, info.type, info.host, keyStrShort)
+
+        // create the replication stream
+        var stream = archive.replicate({live: true})
+        archive.replicationStreams.push(stream)
+        stream.once('close', () => {
+          var rs = archive.replicationStreams
+          var i = rs.indexOf(stream)
+          if (i !== -1) rs.splice(rs.indexOf(stream), 1)
+        })
+
+        // timeout the connection after 5s if handshake does not occur
+        var TO = setTimeout(() => {
+          debug('handshake timeout chan=%s type=%s host=%s key=%s', chan, info.type, info.host, keyStrShort)
+          stream.destroy(new Error('Timed out waiting for handshake'))
+        }, 5000)
+        stream.once('handshake', () => clearTimeout(TO))
+
+        // debugging
+        stream.on('error', err => debug('error chan=%s type=%s host=%s key=%s', chan, info.type, info.host, keyStrShort, err))
+        stream.on('close', () => debug('closing connection chan=%s type=%s host=%s key=%s', chan, info.type, info.host, keyStrShort))
+        return stream
+      }
+    }))
+    swarm.listen(this.config.datPort)
```

Comment:
This will make all the swarms try to listen on the same port, right? Seems like we'll either need a port range (for multiple swarms) or a single swarm like hypercore-archiver.